Mathlib/Geometry/Manifold/VectorBundle/LocalFrame.lean

```diff
@@ -0,0 +1,281 @@
+  {M : Type*} [TopologicalSpace M] [ChartedSpace H M] [IsManifold I 0 M]
+
+variable {F : Type*} [NormedAddCommGroup F] [NormedSpace 𝕜 F]
+  -- `F` model fiber
+  (n : WithTop ℕ∞)
+  {V : M → Type*} [TopologicalSpace (TotalSpace F V)]
+  [∀ x, AddCommGroup (V x)] [∀ x, Module 𝕜 (V x)]
+  [∀ x : M, TopologicalSpace (V x)]
+  -- not needed in this file
+  -- [∀ x, IsTopologicalAddGroup (V x)] [∀ x, ContinuousSMul 𝕜 (V x)]
+  [FiberBundle F V] [VectorBundle 𝕜 F V] [ContMDiffVectorBundle n F V I]
+  -- `V` vector bundle
+
+noncomputable section
+
+section IsLocalFrame
+
+omit [IsManifold I 0 M] [VectorBundle 𝕜 F V]
+
+variable {ι : Type*} {s s' : ι → (x : M) → V x} {u u' : Set M} {x : M} {n : WithTop ℕ∞}
+
+variable (I F n) in
+/--
+A family of sections `s i` of `V → M` is called a **C^k local frame** on a set `U ⊆ M` iff
+- the section values `s i x` form a basis for each `x ∈ U`,
+- each section `s i` is `C^k` on `U`.
+-/
+structure IsLocalFrameOn (s : ι → (x : M) → V x) (u : Set M) where
+  linearIndependent {x : M} (hx : x ∈ u) : LinearIndependent 𝕜 (s · x)
+  generating {x : M} (hx : x ∈ u) : ⊤ ≤ Submodule.span 𝕜 (Set.range (s · x))
+  contMDiffOn (i : ι) : CMDiff[u] n (T% (s i))
+
+namespace IsLocalFrameOn
+
+/-- If `s = s'` on `u` and `s i` is a local frame on `u`, then so is `s'`. -/
+lemma congr (hs : IsLocalFrameOn I F n s u) (hs' : ∀ i, ∀ x, x ∈ u → s i x = s' i x) :
+    IsLocalFrameOn I F n s' u where
+  linearIndependent := by
+    intro x hx
+    have := hs.linearIndependent hx
+    simp_all
+  generating := by
+    intro x hx
+    have := hs.generating hx
+    simp_all
+  contMDiffOn i := (hs.contMDiffOn i).congr fun y hy ↦ by simp [hs' i y hy]
+
+lemma mono (hs : IsLocalFrameOn I F n s u) (hu'u : u' ⊆ u) : IsLocalFrameOn I F n s u' where
+  linearIndependent := by
+    intro x hx
+    exact hs.linearIndependent (hu'u hx)
+  generating := by
+    intro x hx
+    exact hs.generating (hu'u hx)
+  contMDiffOn i := (hs.contMDiffOn i).mono hu'u
+
+lemma contMDiffAt (hs : IsLocalFrameOn I F n s u) (hu : IsOpen u) (hx : x ∈ u) (i : ι) :
+    CMDiffAt n (T% (s i)) x :=
+  (hs.contMDiffOn i).contMDiffAt <| hu.mem_nhds hx
+
+/-- Given a local frame `{s i}` on `U ∋ x`, returns the basis `{s i}` of `V x` -/
+def toBasisAt (hs : IsLocalFrameOn I F n s u) (hx : x ∈ u) : Basis ι 𝕜 (V x) :=
+  Basis.mk (hs.linearIndependent hx) (hs.generating hx)
+
+@[simp]
+lemma toBasisAt_coe (hs : IsLocalFrameOn I F n s u) (hx : x ∈ u) (i : ι) :
+    (toBasisAt hs hx) i = s i x := by
+  simpa only [toBasisAt] using Basis.mk_apply (hs.linearIndependent hx) (hs.generating hx) i
+
+open scoped Classical in
+/-- Coefficients of a section `s` of `V` w.r.t. a local frame `{s i}` on `u`.
+Outside of `u`, this returns the junk value 0. -/
+-- NB. We don't use simps here, as we prefer to have dedicated `_apply` lemmas for the separate
+-- cases.
+def repr (hs : IsLocalFrameOn I F n s u) (i : ι) : (Π x : M, V x) →ₗ[𝕜] M → 𝕜 where
```

Comment:
I have never been a fan of the name `repr` in `Basis`. I'd prefer almost anything else here. What do you think of:
```suggestion
def toLin (hs : IsLocalFrameOn I F n s u) (i : ι) : (Π x : M, V x) →ₗ[𝕜] M → 𝕜 where
```

Reply:
> Why did you pick `toLin`?

I think I was probably vaguely thinking of `Matrix.toLin`.

> To me, that seems "type-correct", but carries little semantic information (unlike coeff or coord).

Agreed!

Reply:
I've gone with `coeff` now